api/osrd_infra/migrations/__init__.py

```diff
@@ -35,9 +50,209 @@ def run_sql_add_foreign_key(model_name: str, field_name: str, link_model: str, n
+    if conflicts := [
+        col for col in itertools.chain(search_columns, extra_columns) if SEARCH_RESERVED_COLUMNS_REGEXP.match(col)
+    ]:
+        raise ValueError(f"reserved column name conflicts: '{conflicts}'")
+    if not SEARCH_TABLE_REGEXP.match(name):
+        raise ValueError(f"invalid table name '{name}', must match {SEARCH_TABLE_REGEXP.pattern}")
+    if conflicts := [
+        col for col in itertools.chain(search_columns, extra_columns) if not SEARCH_COLUMN_REGEXP.match(name)
+    ]:
+        raise ValueError(f"invalid column names '{conflicts}', must match {SEARCH_COLUMN_REGEXP.pattern}")
+    # Defaults to the view's name camel case-d
+    phony_model_name = phony_model_name or name.replace("_", " ").replace("-", " ").title().replace(" ", "")
+    create_indexes = ";\n".join(index.create(col, name) for col in search_columns)
+    column_exprs = {col: f"osrd_prepare_for_search(({sql}))" for col, sql in search_columns.items()} | {
+        col: sql for col, (sql, _) in extra_columns.items()
+    }
+    select_columns = ", ".join(
+        itertools.chain(
+            [f'"{source_table}"."{source_table_pk}" AS id'],
+            (f'({sql}) AS "{col}"' for col, sql in column_exprs.items()),
+        )
+    )
+    select = f"""SELECT {select_columns} FROM "{source_table}" {query_continuation}"""
+    if table_type is SearchTableType.TABLE:
+        # NOTE: here source_table has to be a physical table and not a subquery
+        search_columns = ", ".join(
+            f'"{col}" {type_}'
+            for col, type_ in itertools.chain(
+                ((col, "TEXT") for col in search_columns), ((col, type_) for col, (_, type_) in extra_columns.items())
+            )
+        )
+        sep = "," if create_table_continuation else ""
+        create_table = f"""CREATE TABLE "{name}" (
+            id BIGINT REFERENCES "{source_table}"("{source_table_pk}") ON UPDATE CASCADE ON DELETE CASCADE,
+            {search_columns}{sep}
+            {create_table_continuation}
+        )""".format(
+            source=source_table
+        )
+        insert_table = f"""INSERT INTO "{name}" {select}""".format(source=source_table)
+        create_triggers, drop_triggers = "", ""
+        if triggers:
+            c_ins, d_ins = insert_trigger(name, source_table, column_exprs, query_continuation)
+            c_upd, d_upd = update_trigger(name, source_table, column_exprs, query_continuation)
+            create_triggers = f"{c_ins}\n{c_upd}"
+            drop_triggers = f"{d_ins}\n{d_upd}"
+        sql = f"{create_table};\n{create_indexes};\n{create_triggers};\n{insert_table}"
+        reverse_sql = f'DROP TABLE "{name}";\n{drop_triggers}'
+    else:
+        table = "MATERIALIZED VIEW" if table_type is SearchTableType.MATERIALIZED_VIEW else "VIEW"
+        table_sql = f"""CREATE {table} "{name}" AS ({select})""".format(source=source_table)
+        sql = f"{table_sql};\n{create_indexes}"
+        reverse_sql = f'DROP {table} "{name}";'
```

Comment:
```suggestion
    sql: str
    reverse_sql: str
    if table_type is SearchTableType.TABLE:
        sql, reverse_sql = _create_sql_migration_script_for_search_table(...)
    else:
        sql, reverse_sql = _create_sql_migration_script_for_materialized_view(...)
```
I think we could declare and split this in private methods to decrease complexity

Reply:
I agree on the principle but these statements depend on many variable defined above and the split would look like this:

```python
sql, reverse_sql = _create_sql_for_table(
            name,
            source_table,
            source_table_pk,
            search_columns,
            extra_columns,
            column_exprs,
            triggers,
            create_indexes,
            query_continuation,
            create_table_continuation,
            select,
        )
```

Ideally we would construct a builder but since we're likely to drop api the next PI I'm not sure it's worth the effort.